### ISSUE TITLE
Seems there's memory issue in pytorch 1.0.0

### ISSUE THREAD

**Issue by zzj0311** [pytorch/pytorch#15774]
## 🐛 Bug
My code works fine under pytorch 0.4.1 (around 3G for now as I'm still running test) while it 'eat up' all my memory (200+G at most) gradually under pytorch 1.0.0. Really have no idea what was happened for now... The particular env is listed follow

Be more specific, these are two dockers in the same physical host and the code is tested under the same graphic card, using around 10G graphic mem.

## Environment

The problem one
```
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla K80
GPU 1: Tesla K80
GPU 2: Tesla K80
GPU 3: Tesla K80
GPU 4: Tesla K80
GPU 5: Tesla K80
GPU 6: Tesla K80
GPU 7: Tesla K80

Nvidia driver version: 410.79
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.4
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a
/usr/local/lib/python3.6/dist-packages/cntk/libs/libcudnn.so.7

Versions of relevant libraries:
[pip] numpy (1.15.4)
[pip] numpydoc (0.7.0)
[pip] torch (1.0.0)
[pip] torchvision (0.2.1)
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.1                      144
[conda] mkl-service               1.1.2            py37h90e4bf4_5
[conda] mkl_fft                   1.0.6            py36hd81dba3_0
[conda] mkl_random                1.0.2            py36hd81dba3_0
[conda] pytorch                   1.0.0           py3.6_cuda9.0.176_cudnn7.4.1_1    pytorch
[conda] torchvision               0.2.1                      py_2    pytorch
```
The normal one
```
PyTorch version: 0.4.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla K80
GPU 1: Tesla K80
GPU 2: Tesla K80
GPU 3: Tesla K80
GPU 4: Tesla K80
GPU 5: Tesla K80
GPU 6: Tesla K80
GPU 7: Tesla K80

Nvidia driver version: 410.79
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.4
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a
/usr/local/lib/python3.6/dist-packages/cntk/libs/libcudnn.so.7

Versions of relevant libraries:
[pip] numpy (1.14.3)
[pip] numpydoc (0.8.0)
[pip] torch (0.4.1.post2)
[pip] torchvision (0.2.1)
[conda] blas                      1.0                         mkl
[conda] mkl                       2018.0.2                      1
[conda] mkl-service               1.1.2            py36h17a0993_4
[conda] mkl_fft                   1.0.1            py36h3010b51_0
[conda] mkl_random                1.0.1            py36h629b387_0
[conda] pytorch                   0.4.1           py36_py35_py27__9.0.176_7.1.2_2    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch
```

**Comment by zzj0311** [comment]
@ailzhang I'll provide it in two days.

**Comment by zzj0311** [comment]
@ailzhang You can check https://github.com/zzj0311/pytorch-image-inpainting-using-mixed-convolution .  Dataset should be placed at the same dir of code, for now. If you need a simplified version, just let me know :)

**Comment by ailzhang** [comment]
@zzj0311 I managed to get your code running on both `0.1` and `0.4.0`, in both cases the program used around `10G` GPU memory and `3G` CPU memory which is very close to the expected behavior.  I cannot repro your memory issue though.

**Comment by zzj0311** [comment]
@ailzhang Initially，yes. But even for 0.4.1, it would take around 20g mem after 100 epoches (it's very time consuming...). For 1.0.0, it ate up all my mem after around 30 epoches while around 60g within 10 epoches (takes about 5 hours though). The graphic mem won't increase significantly if you ask.

**Comment by ailzhang** [comment]
Aha, could you get us a simpler(quicker) repro? Even this is reproducible on our end, it will be very hard to iterate and fix. :( Thanks!

**Comment by zzj0311** [comment]
Oh, one more thing. The problem occurred when I fine tune the model. As I remember it seems not happen when I simply run the model. Use -r [ckpt](https://pan.baidu.com/s/1YTK46whMRGBLAPrZN0PCaA) to resume and -f to enable fine tune. If still can't repro, just forget about it.

**Comment by zzj0311** [comment]
Hmmm, I'll simplify it then.

**Comment by zzj0311** [comment]
@ailzhang Hi, I just reproduced the bug with memory usage info logged and it occurred in the second epoch!
```
Fri, 18 Jan 2019 12:07:04 modelTrain.py[line:59] : using device:cuda:4
Fri, 18 Jan 2019 12:07:08 util.py[line:172] : -----------epoch 130 start here----------
Fri, 18 Jan 2019 12:07:29 util.py[line:194] : Mem: 2.137744903564453 G
Fri, 18 Jan 2019 12:08:45 util.py[line:194] : Mem: 2.2248687744140625 G
Fri, 18 Jan 2019 12:09:56 util.py[line:194] : Mem: 2.2288360595703125 G
Fri, 18 Jan 2019 12:11:09 util.py[line:194] : Mem: 2.228862762451172 G
Fri, 18 Jan 2019 12:12:22 util.py[line:194] : Mem: 2.228893280029297 G
Fri, 18 Jan 2019 12:13:34 util.py[line:194] : Mem: 2.2333106994628906 G
Fri, 18 Jan 2019 12:14:46 util.py[line:194] : Mem: 2.2333602905273438 G
Fri, 18 Jan 2019 12:15:58 util.py[line:194] : Mem: 2.2333908081054688 G
Fri, 18 Jan 2019 12:17:10 util.py[line:194] : Mem: 2.2334213256835938 G
Fri, 18 Jan 2019 12:18:22 util.py[line:194] : Mem: 2.233448028564453 G
Fri, 18 Jan 2019 12:19:34 util.py[line:194] : Mem: 2.233478546142578 G
Fri, 18 Jan 2019 12:20:47 util.py[line:194] : Mem: 2.2335052490234375 G
Fri, 18 Jan 2019 12:21:58 util.py[line:194] : Mem: 2.233531951904297 G
Fri, 18 Jan 2019 12:23:09 util.py[line:194] : Mem: 2.233562469482422 G
Fri, 18 Jan 2019 12:24:19 util.py[line:194] : Mem: 2.2335891723632812 G
Fri, 18 Jan 2019 12:25:33 util.py[line:194] : Mem: 2.245494842529297 G
Fri, 18 Jan 2019 12:26:44 util.py[line:194] : Mem: 2.254467010498047 G
Fri, 18 Jan 2019 12:27:56 util.py[line:194] : Mem: 2.2544937133789062 G
Fri, 18 Jan 2019 12:29:08 util.py[line:194] : Mem: 2.2545242309570312 G
Fri, 18 Jan 2019 12:30:20 util.py[line:194] : Mem: 2.2604103088378906 G
Fri, 18 Jan 2019 12:31:32 util.py[line:194] : Mem: 2.2604408264160156 G
Fri, 18 Jan 2019 12:32:44 util.py[line:194] : Mem: 2.2455101013183594 G
Fri, 18 Jan 2019 12:33:57 util.py[line:194] : Mem: 2.2689781188964844 G
Fri, 18 Jan 2019 12:34:46 util.py[line:172] : -----------epoch 131 start here----------
Fri, 18 Jan 2019 12:34:48 util.py[line:194] : Mem: 2.234161376953125 G
Fri, 18 Jan 2019 12:35:56 util.py[line:194] : Mem: 2.5387840270996094 G
Fri, 18 Jan 2019 12:37:04 util.py[line:194] : Mem: 2.8317527770996094 G
Fri, 18 Jan 2019 12:38:12 util.py[line:194] : Mem: 3.1247215270996094 G
Fri, 18 Jan 2019 12:39:21 util.py[line:194] : Mem: 3.4176902770996094 G
Fri, 18 Jan 2019 12:40:29 util.py[line:194] : Mem: 3.7106590270996094 G
Fri, 18 Jan 2019 12:41:38 util.py[line:194] : Mem: 3.9919090270996094 G
Fri, 18 Jan 2019 12:42:46 util.py[line:194] : Mem: 4.284908294677734 G
Fri, 18 Jan 2019 12:43:55 util.py[line:194] : Mem: 4.577903747558594 G
Fri, 18 Jan 2019 12:45:03 util.py[line:194] : Mem: 4.859180450439453 G
```
While it still takes half an hour to get to the second epoch... Maybe I could reduce the training data?

**Comment by zzj0311** [comment]
@ailzhang I may no longer be able to test the code after Friday cst as the course is over and I'll have no auth to use the cluster after that. Please let me know if there's still a thing I can help before that date, much THX!

**Comment by ailzhang** [comment]
@zzj0311 Could you copy the `ckpt` to `dropbox` or `google drive`? I cannot download the file without installing the app and opening a new account. 
Thanks!

**Comment by zzj0311** [comment]
@ailzhang You can now get it [here](https://www.dropbox.com/sh/ds5s5016kf42bjj/AABy4iLua26Jp06egZGSuNWja?dl=0)

**Comment by otaj** [comment]
I don't know was exactly would be the issue, but I experienced something somewhat similar, which I actually managed to pinpoint. The reason for my memory increase was using DataLoader with custom Dataset (subclassing from torch.utils.data.Dataset) where I was storing loading data in a dict instead of a simple list to accommodate shuffling. However, the indices going to method __getitem__ (generated by DataLoader) were not simple integers as I anticipated (and as the doc would suggest, https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset) but 0-D dimensional integer tensors. And since these don't have the __hash__ depending on a value, rather on memory location, the dict would grow and grow and grow, because it didn't know there was something in it already. Hope I made it somewhat clear. To illustrate it more:
`class LiDARDataset(torch.utils.data.Dataset):
    def __init__(self, folder, keep_ram=False):
        self.names = list(filter(self.shape_check, sorted(glob.glob(os.path.join(folder, '*.npy')))))
        self.keep_ram = keep_ram
        if self.keep_ram:
            self.loaded = dict()
        else:
            self.loaded = weakref.WeakValueDictionary()

    def __len__(self) -> int:
        return len(self.names)

    def __getitem__(self, key: int) -> BatchType:
        if not isinstance(key, int): # without these two lines, it would grow
            key = key.item() 
        result = self.loaded.get(key) # because the result would always be None
        if result is None:
            result = self.transform(np.load(self.names[key]), key)
            self.loaded[key] = result
        return result`

So, check you don't have any dictionary, which is indexed by torch.tensor, because that is what is changed since 0.4.1. Or you don't rely on hashing of torch.tensor by its value rather than the memory location

**Comment by zzj0311** [comment]
@otaj Thx but no, I would never do that. Shuffling a dict is a quiet dumb operation

**Comment by otaj** [comment]
Shuffling a dict is a quite dumb operation, but I am not saying you did that - I am saying you might index a dict by instances of torch.tensor when you thought you were working with numbers.

**Comment by zzj0311** [comment]
@otaj hmmm but still not my case. I'm pretty sure I was using list of file names as my dataset so the problem most likely to be something related to multi-threads image loading process. Thanks for your time.

**Comment by otaj** [comment]
So, random_split maybe? See #17466

**Comment by mruberry** [comment]
Closing due to age. If this is still an issue on PyTorch 1.7 or later would you please file a new issue?
